### PR TITLE
Caching and simple tests

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/managers/pip.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export default class extends Manager {
     }
 
     async getMeta(name: string, version?: string): Promise<Meta> {
-        const json = await this.requestPIP(name, version);
+        const json = await this.cache(this.requestPIP, name, version);
         return {
             package: `${json.name}:${json.version}`,
```
Comment (xferra) [inline src/managers/pip.ts]: Do not change the original name.

File: test/npm/test.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+const {expect} = require('chai');
+const {getManager, managerType} = require('../../dist/index');
+
+const manager = getManager(managerType.npm);
+
+describe('Npm', function() {
+  describe('getMeta()', function() {
+    const package = 'eslint';
+    const version = '5.3.0';
+    it(`should return meta of ${package} v${version}`, function(done) {
```
Comment (xferra) [inline test/npm/test.js]: Do not mix arrow funcs with ES6 syntax.

File: test/npm/test.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+const {expect} = require('chai');
+const {getManager, managerType} = require('../../dist/index');
```
Comment (xferra) [inline test/npm/test.js]: Test dist not src?

File: src/managers/gem.ts [hunk 4]
```diff
@@ -34,7 +33,7 @@ export default class extends Manager {
     }
 
     async getDeps(name: string, version?: string): Promise<Dependency[]> {
-        const json = await this.requestGEM(name, version);
+        const json = await this.cache(this.requestGEM, name, version);
         const dependencies: Dependency[] = [];
```
Comment (xferra) [inline src/managers/gem.ts]: Try avoid foreach use filter+map instead.

File: src/managers/npm.ts [hunk 5]
```diff
@@ -6,12 +6,12 @@ import packageJson from 'package-json';
 export default class extends Manager {
 
     // TODO: Create filters to parse only necessary data
-    private requestNPM(name: string, version?: string): Promise<any> {
-        return packageJson(name, { version: version ? version : 'latest', fullMetadata: true, allVersions: true });
+    private async requestNPM(name: string, version?: string): Promise<any> {
+        return await packageJson(name, { version: version ? version : 'latest', fullMetadata: true, allVersions: true });
```
Comment (xferra) [inline src/managers/npm.ts]: Use the factory pattern.

File: src/interface/manager.ts [hunk 6]
```diff
@@ -7,9 +7,31 @@ import { Meta } from './meta';
 export abstract class Manager {
 
     protected name: string;
+    private cached: { [key: string]: any };
```
Comment (xferra) [inline src/interface/manager.ts]: Isolate in a separate class.

File: src/index.ts [hunk 7]
```diff
@@ -9,8 +9,11 @@ export enum managerType {
...
 
 
-export function getManager(type: managerType): Manager {
-    const managerName: string = managerType[type];
+export function getManager(manager: managerType | String): Manager {
+    const managerName: string =
+        manager instanceof (String) ?
+            manager.toString() :
+            managerType[manager];
     const managerClass = require(`./managers/${managerName}`);
```
Comment (xferra) [inline src/index.ts]: Not secure, include all managers at design-time.

File: src/interface/manager.ts [hunk 8]
```diff
@@ -7,9 +7,31 @@ import { Meta } from './meta';
...
 
     constructor(name: string) {
         this.name = name;
```
Comment (xferra) [inline src/interface/manager.ts]: The name seems to e useless in this context.

File: src/interface/manager.ts [hunk 9]
```diff
@@ -7,9 +7,31 @@ import { Meta } from './meta';
...
+        this.cached[key] = value;
+    }
+
+    protected async cache(
+        request: (name: string, version?: string) => Promise<any>, 
+        name: string, 
+        version?: string ) : Promise<any> 
+        {
+        const key = `${name}:${version}`;
+        if(!this.requestCache(key)) {
```
Comment (xferra) [inline src/interface/manager.ts]: Ha! It's possible that you will start several promises before 1st result will be saved.

File: test/npm/test.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+const manager = getManager(managerType.npm);
```
Comment (xferra) [inline test/npm/test.js]: Put it into a separate folder, it's integration, not a unit test.

File: test/npm/test.js (hunk 3 above)
Comment (ghost) [reply test/npm/test.js]: Dist folder contains builded js files

File: src/interface/manager.ts (hunk 8 above)
Comment (ghost) [reply src/interface/manager.ts]: We need name property in `manager` classes for deps generation

Comment (xferra) [reply src/interface/manager.ts]: No, we don't. I think it's obvious that deps from such class depend on the same pkg man.

File: src/index.ts [hunk 11]
```diff
@@ -5,12 +5,18 @@ export enum managerType {
...
-    const managerClass = require(`./managers/${managerName}`);
-    return new managerClass.default(managerName);
-}
+export function getManager(manager: managerType | String): Manager | undefined {
+    const managerName: string | undefined =
+        manager instanceof (String) ?
+            ((Object as any).values(managerType).includes(manager.toString()) ? manager.toString() : undefined) :
+            managerType[manager];
+    if (managerName){
+        const managerClass = require(`./managers/${managerName}`);
```
Comment (xferra) [inline src/index.ts:17]: Instead of requesting/including js file y name I strongly recommend to define them explicitly.

Comment (xferra) [reply src/index.ts:17]: + if the enum is defined why do you overcomplicate the logic that checks the string (manager)

Comment (xferra) [reply src/index.ts:17]: And it's actually registry...

**Comment by romanrostislavovich** [comment]
@doroxz ping

**Comment by doroxz-zz** [comment]
Will be added https://github.com/linterhub/registry/pull/33

**Pull request closed without merging**